File: src/java/magmac/app/lang/node/Arguments.java
package magmac.app.lang.node;

import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.NodeListRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.Rule;
import magmac.app.lang.Deserializers;
import magmac.app.lang.java.JavaDeserializers;
import magmac.app.lang.JavaRules;
import magmac.app.lang.java.JavaLang;

public final class Arguments {
    public static CompileResult<JavaLang.JavaArgument> deserialize(Node node) {
        return Deserializers.orError("argument", node, Lists.of(
                Deserializers.wrap(JavaDeserializers::deserializeWhitespace),
                Deserializers.wrap(JavaDeserializers::deserializeValue)
        ));
    }

    public static Rule createArgumentsRule(Rule value) {
        return NodeListRule.Values("arguments", new OrRule(Lists.of(
                JavaRules.createWhitespaceRule(),
                value
        )));
    }
}

File: src/web/magmac/app/lang/node/Arguments.ts
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { NodeListRule } from "../../../../magmac/app/compile/rule/NodeListRule";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { Deserializers } from "../../../../magmac/app/lang/Deserializers";
import { JavaDeserializers } from "../../../../magmac/app/lang/java/JavaDeserializers";
import { JavaRules } from "../../../../magmac/app/lang/JavaRules";
import { JavaLang } from "../../../../magmac/app/lang/java/JavaLang";
export class Arguments {
	deserialize(node : Node) : CompileResult<JavaLang.JavaArgument> {return 0.orError( 0, 0, 0.of( 0.wrap( 0.deserializeWhitespace), 0.wrap( 0.deserializeValue)));;}
	createArgumentsRule(value : Rule) : Rule {return 0.Values( 0, new OrRule( 0.of( 0.createWhitespaceRule( ), 0)));;}
}
